File: cmd/podium-mcp/load_domain_test.go
package main

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/lennylabs/podium/pkg/manifest"
	"github.com/lennylabs/podium/pkg/registry/filesystem"
)

// fakeRegistry serves /v1/load_domain (the supplied payload) and /v1/catalog
// (the supplied artifacts, filtered to the requested scope prefix), the two
// endpoints the client-side load_domain merge consults.
func fakeRegistry(t *testing.T, loadDomain map[string]any, catalog []catalogEntry) *mcpServer {
	t.Helper()
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/v1/load_domain":
			_ = json.NewEncoder(w).Encode(loadDomain)
		case "/v1/catalog":
			scope := r.URL.Query().Get("scope")
			ids := []string{}
			arts := []catalogEntry{}
			for _, e := range catalog {
				if scope == "" || e.ID == scope || hasSegPrefix(e.ID, scope) {
					ids = append(ids, e.ID)
					arts = append(arts, e)
				}
			}
			_ = json.NewEncoder(w).Encode(map[string]any{"ids": ids, "artifacts": arts})
		default:
			http.Error(w, "not found", http.StatusNotFound)
		}
	}))
	t.Cleanup(ts.Close)
	return &mcpServer{cfg: &config{registry: ts.URL}, http: &http.Client{}}
}

func hasSegPrefix(id, prefix string) bool {
	return len(id) > len(prefix) && id[:len(prefix)] == prefix && id[len(prefix)] == '/'
}

// decodeDomain runs the merge result (a jsonAny map) back through JSON into the
// typed envelope so tests assert against the wire schema.
func decodeDomain(t *testing.T, out any) ldResponse {
	t.Helper()
	b, err := json.Marshal(out)
	if err != nil {
		t.Fatalf("marshal result: %v", err)
	}
	var resp ldResponse
	if err := json.Unmarshal(b, &resp); err != nil {
		t.Fatalf("decode result %s: %v", b, err)
	}
	return resp
}

func notableIDs(resp ldResponse) map[string]ldArtifact {
	m := map[string]ldArtifact{}
	for _, a := range resp.Notable {
		m[a.ID] = a
	}
	return m
}

func subdomainByPath(resp ldResponse, path string) (ldSubdomain, bool) {
	for _, s := range resp.Subdomains {
		if s.Path == path {
			return s, true
		}
	}
	return ldSubdomain{}, false
}

// spec: §4.5.2/§4.5.4/§6.4 (F-4.5.2/F-6.4.2) — with no workspace overlay the
// load_domain result passes through the registry untouched.
func TestLoadDomain_NoOverlayPassthrough(t *testing.T) {
	t.Parallel()
	srv := fakeRegistry(t, map[string]any{
		"path":        "finance",
		"description": "Registry finance",
		"keywords":    []string{"money"},
		"subdomains":  []map[string]any{{"path": "finance/ap", "name": "ap", "description": "AP"}},
		"notable":     []map[string]any{{"id": "finance/x", "type": "skill", "summary": "x", "source": "signal"}},
	}, nil)
	resp := decodeDomain(t, srv.loadDomain(map[string]any{"path": "finance"}))
	if resp.Description != "Registry finance" {
		t.Errorf("description = %q, want registry value", resp.Description)
	}
	if len(resp.Notable) != 1 || resp.Notable[0].ID != "finance/x" {
		t.Errorf("notable = %+v, want only finance/x", resp.Notable)
	}
	if len(resp.Subdomains) != 1 || resp.Subdomains[0].Path != "finance/ap" {
		t.Errorf("subdomains = %+v, want only finance/ap", resp.Subdomains)
	}
}

// spec: §4.5.4 — an overlay DOMAIN.md description/body wins over the registry's
// (highest-precedence layer) and keywords append-unique.
func TestLoadDomain_OverlayDescriptionAndKeywords(t *testing.T) {
	t.Parallel()
	srv := fakeRegistry(t, map[string]any{
		"path":        "finance",
		"description": "Registry finance",
		"keywords":    []string{"money", "ledger"},
		"subdomains":  []map[string]any{},
		"notable":     []map[string]any{},
	}, nil)
	srv.overlayDomains = map[string]*manifest.Domain{
		"finance": {
			Body:      "Local working notes for finance",
			Discovery: &manifest.DomainDiscovery{Keywords: []string{"ledger", "draft"}},
		},
	}
	resp := decodeDomain(t, srv.loadDomain(map[string]any{"path": "finance"}))
	if resp.Description != "Local working notes for finance" {
		t.Errorf("description = %q, want overlay body", resp.Description)
	}
	want := []string{"money", "ledger", "draft"}
	if len(resp.Keywords) != len(want) {
		t.Fatalf("keywords = %v, want %v", resp.Keywords, want)
	}
	for i := range want {
		if resp.Keywords[i] != want[i] {
			t.Errorf("keywords = %v, want %v", resp.Keywords, want)
		}
	}
}

// spec: §4.5.5 — an overlay artifact that is a direct child of the requested
// domain joins the notable candidate pool, tagged overlay-sourced.
func TestLoadDomain_OverlayDirectChildNotable(t *testing.T) {
	t.Parallel()
	srv := fakeRegistry(t, map[string]any{
		"path":       "finance",
		"subdomains": []map[string]any{},
		"notable":    []map[string]any{{"id": "finance/x", "type": "skill", "summary": "x", "source": "signal"}},
	}, nil)
	srv.overlay = []filesystem.ArtifactRecord{{
		ID: "finance/draft-helper",
		Artifact: &manifest.Artifact{
			Type: manifest.TypeSkill, Name: "draft-helper", Version: "0.1.0",
			Description: "in-progress finance helper",
		},
	}}
	resp := decodeDomain(t, srv.loadDomain(map[string]any{"path": "finance"}))
	got := notableIDs(resp)
	d, ok := got["finance/draft-helper"]
	if !ok {
		t.Fatalf("overlay draft missing from notable: %+v", resp.Notable)
	}
	if !d.Overlay || d.Summary != "in-progress finance helper" || d.Source != "signal" {
		t.Errorf("overlay notable = %+v, want overlay/signal with summary", d)
	}
	if _, ok := got["finance/x"]; !ok {
		t.Errorf("registry notable finance/x dropped: %+v", resp.Notable)
	}
}

// spec: §4.5.5 — an overlay artifact below an immediate child introduces that
// child as a subdomain of the requested domain.
func TestLoadDomain_OverlayNewSubdomain(t *testing.T) {
	t.Parallel()
	srv := fakeRegistry(t, map[string]any{
		"path":       "finance",
		"subdomains": []map[string]any{{"path": "finance/ap", "name": "ap", "description": "AP"}},
		"notable":    []map[string]any{},
	}, nil)
	srv.overlay = []filesystem.ArtifactRecord{{
		ID:       "finance/newteam/draft",
		Artifact: &manifest.Artifact{Type: manifest.TypeSkill, Name: "draft", Version: "0.1.0"},
	}}
	srv.overlayDomains = map[string]*manifest.Domain{
		"finance/newteam": {Description: "New team workspace"},
	}
	resp := decodeDomain(t, srv.loadDomain(map[string]any{"path": "finance"}))
	sd, ok := subdomainByPath(resp, "finance/newteam")
	if !ok {
		t.Fatalf("overlay subdomain finance/newteam missing: %+v", resp.Subdomains)
	}
	if sd.Name != "newteam" || sd.Description != "New team workspace" {
		t.Errorf("subdomain = %+v, want name/desc from overlay", sd)
	}
	if _, ok := subdomainByPath(resp, "finance/ap"); !ok {
		t.Errorf("registry subdomain finance/ap dropped: %+v", resp.Subdomains)
	}
}

// spec: §4.5.2 — a workspace-local DOMAIN.md include: resolves over the merged
// view (registry catalog ∪ overlay), pulling in both a registry artifact and an
// overlay artifact.
func TestLoadDomain_OverlayIncludeMergedView(t *testing.T) {
	t.Parallel()
	srv := fakeRegistry(t, map[string]any{
		"path":       "drafts",
		"subdomains": []map[string]any{},
		"notable":    []map[string]any{},
	}, []catalogEntry{
		{ID: "finance/ap/registry-pay", Type: "skill", Summary: "registry pay"},
		{ID: "other/unrelated", Type: "skill", Summary: "nope"},
	})
	srv.overlay = []filesystem.ArtifactRecord{{
		ID:       "finance/ap/overlay-pay",
		Artifact: &manifest.Artifact{Type: manifest.TypeSkill, Name: "overlay-pay", Version: "0.1.0", Description: "overlay pay"},
	}}
	srv.overlayDomains = map[string]*manifest.Domain{
		"drafts": {Include: []string{"finance/ap/*"}},
	}
	resp := decodeDomain(t, srv.loadDomain(map[string]any{"path": "drafts"}))
	got := notableIDs(resp)
	if _, ok := got["finance/ap/registry-pay"]; !ok {
		t.Errorf("merged-view include missed registry artifact: %+v", resp.Notable)
	}
	if _, ok := got["finance/ap/overlay-pay"]; !ok {
		t.Errorf("merged-view include missed overlay artifact: %+v", resp.Notable)
	}
	if _, ok := got["other/unrelated"]; ok {
		t.Errorf("include pulled in out-of-scope artifact: %+v", resp.Notable)
	}
}

// spec: §4.5.3 — an overlay DOMAIN.md unlisted: true removes the folder and its
// subtree from the parent's enumeration.
func TestLoadDomain_OverlayUnlistedPrunesSubdomain(t *testing.T) {
	t.Parallel()
	srv := fakeRegistry(t, map[string]any{
		"path": "finance",
		"subdomains": []map[string]any{
			{"path": "finance/ap", "name": "ap", "description": "AP"},
			{"path": "finance/secret", "name": "secret", "description": "Secret"},
		},
		"notable": []map[string]any{},
	}, nil)
	srv.overlayDomains = map[string]*manifest.Domain{
		"finance/secret": {Unlisted: true},
	}
	resp := decodeDomain(t, srv.loadDomain(map[string]any{"path": "finance"}))
	if _, ok := subdomainByPath(resp, "finance/secret"); ok {
		t.Errorf("unlisted subdomain not pruned: %+v", resp.Subdomains)
	}
	if _, ok := subdomainByPath(resp, "finance/ap"); !ok {
		t.Errorf("non-unlisted subdomain finance/ap dropped: %+v", resp.Subdomains)
	}
}

// spec: §4.5.5 — an overlay DOMAIN.md at a child path overrides that child's
// short description (frontmatter description, never the body).
func TestLoadDomain_OverlayChildDescriptionOverride(t *testing.T) {
	t.Parallel()
	srv := fakeRegistry(t, map[string]any{
		"path":       "finance",
		"subdomains": []map[string]any{{"path": "finance/ap", "name": "ap", "description": "Registry AP"}},
		"notable":    []map[string]any{},
	}, nil)
	srv.overlayDomains = map[string]*manifest.Domain{
		"finance/ap": {Description: "Local AP overrides", Body: "body never shown for a child"},
	}
	resp := decodeDomain(t, srv.loadDomain(map[string]any{"path": "finance"}))
	sd, ok := subdomainByPath(resp, "finance/ap")
	if !ok {
		t.Fatalf("finance/ap subdomain missing")
	}
	if sd.Description != "Local AP overrides" {
		t.Errorf("child description = %q, want overlay frontmatter description", sd.Description)
	}
}

File: sdks/podium-ts/src/load_domain.test.ts
// Client.loadDomain workspace-overlay merge (F-4.5.2 / F-6.4.2). The merge
// mirrors the Go MCP bridge (cmd/podium-mcp/load_domain.go): the overlay is the
// highest-precedence layer in the caller's effective view (§6.4), composed onto
// the registry's rendered load_domain result client-side (§4.5.4, §4.5.5).

import { mkdtemp, mkdir, writeFile, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { Client, type LoadDomainNotable, type LoadDomainResult } from "./index.js";

// overlayArtifact writes one overlay ARTIFACT.md package at the canonical id.
async function overlayArtifact(
  root: string,
  id: string,
  opts: { type?: string; desc?: string; body?: string } = {},
): Promise<void> {
  const pkg = join(root, ...id.split("/"));
  await mkdir(pkg, { recursive: true });
  const fm = `---\ntype: ${opts.type ?? "skill"}\nversion: 0.1.0\ndescription: ${opts.desc ?? ""}\n---\n${opts.body ?? "body"}\n`;
  await writeFile(join(pkg, "ARTIFACT.md"), fm);
}

// overlayDomain writes one DOMAIN.md at a domain path ("" for the overlay root).
async function overlayDomain(root: string, domainPath: string, frontmatterBody: string): Promise<void> {
  const dir = domainPath === "" ? root : join(root, ...domainPath.split("/"));
  await mkdir(dir, { recursive: true });
  await writeFile(join(dir, "DOMAIN.md"), frontmatterBody);
}

function notableByID(resp: LoadDomainResult): Map<string, LoadDomainNotable> {
  const m = new Map<string, LoadDomainNotable>();
  for (const a of resp.notable) m.set(a.id, a);
  return m;
}

function subdomainByPath(resp: LoadDomainResult, path: string): LoadDomainResult["subdomains"][number] | undefined {
  return resp.subdomains.find((s) => s.path === path);
}

// hasSegPrefix reports whether id sits strictly under the scope prefix
// (segment-aligned), mirroring the Go fake registry's catalog scope filter.
function hasSegPrefix(id: string, prefix: string): boolean {
  return id.length > prefix.length && id.startsWith(prefix) && id[prefix.length] === "/";
}

// registryFetcher serves /v1/load_domain (the supplied payload) and /v1/catalog
// (the supplied artifacts, filtered to the requested scope prefix), the two
// endpoints the merge consults.
function registryFetcher(
  loadDomain: Record<string, unknown>,
  catalog: Array<{ id: string; type?: string; summary?: string }> = [],
): typeof fetch {
  return async (input) => {
    const url = new URL(String(input));
    if (url.pathname === "/v1/load_domain") {
      return new Response(JSON.stringify(loadDomain), { status: 200 });
    }
    if (url.pathname === "/v1/catalog") {
      const scope = url.searchParams.get("scope") ?? "";
      const arts = catalog.filter(
        (e) => scope === "" || e.id === scope || hasSegPrefix(e.id, scope),
      );
      return new Response(JSON.stringify({ ids: arts.map((a) => a.id), artifacts: arts }), {
        status: 200,
      });
    }
    return new Response("not found", { status: 404 });
  };
}

describe("loadDomain overlay merge (F-4.5.2/F-6.4.2)", () => {
  let dir: string;
  let overlay: string;
  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "podium-ld-"));
    overlay = join(dir, "overlay");
    await mkdir(overlay, { recursive: true });
  });
  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  // spec: §4.5.2/§4.5.4/§6.4 — with no workspace overlay the load_domain result
  // passes through the registry untouched.
  it("passes through unchanged with no overlay", async () => {
    const fetcher = registryFetcher({
      path: "finance",
      description: "Registry finance",
      keywords: ["money"],
      subdomains: [{ path: "finance/ap", name: "ap", description: "AP" }],
      notable: [{ id: "finance/x", type: "skill", summary: "x", source: "signal" }],
    });
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("finance");
    expect(resp.description).toBe("Registry finance");
    expect(resp.notable.map((n) => n.id)).toEqual(["finance/x"]);
    expect(resp.subdomains.map((s) => s.path)).toEqual(["finance/ap"]);
  });

  // spec: §4.5.4 — an overlay DOMAIN.md body wins over the registry description
  // (highest-precedence layer) and keywords append-unique.
  it("merges overlay description body and keywords", async () => {
    await overlayDomain(
      overlay,
      "finance",
      "---\ndiscovery:\n  keywords: [ledger, draft]\n---\nLocal working notes for finance\n",
    );
    const fetcher = registryFetcher({
      path: "finance",
      description: "Registry finance",
      keywords: ["money", "ledger"],
      subdomains: [],
      notable: [],
    });
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("finance");
    expect(resp.description?.trim()).toBe("Local working notes for finance");
    expect(resp.keywords).toEqual(["money", "ledger", "draft"]);
  });

  // spec: §4.5.5 — an overlay artifact that is a direct child of the requested
  // domain joins the notable candidate pool, tagged overlay-sourced; the
  // registry notable entry is retained.
  it("surfaces a direct-child overlay artifact as notable", async () => {
    await overlayArtifact(overlay, "finance/draft-helper", { desc: "in-progress finance helper" });
    const fetcher = registryFetcher({
      path: "finance",
      subdomains: [],
      notable: [{ id: "finance/x", type: "skill", summary: "x", source: "signal" }],
    });
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("finance");
    const got = notableByID(resp);
    const d = got.get("finance/draft-helper");
    expect(d).toBeDefined();
    expect(d?.overlay).toBe(true);
    expect(d?.source).toBe("signal");
    expect(d?.summary).toBe("in-progress finance helper");
    expect(got.has("finance/x")).toBe(true);
  });

  // spec: §4.5.5 — an overlay artifact below an immediate child introduces that
  // child as a subdomain of the requested domain; the registry subdomain stays.
  it("introduces an overlay-only subdomain", async () => {
    await overlayArtifact(overlay, "finance/newteam/draft");
    await overlayDomain(overlay, "finance/newteam", "---\ndescription: New team workspace\n---\n");
    const fetcher = registryFetcher({
      path: "finance",
      subdomains: [{ path: "finance/ap", name: "ap", description: "AP" }],
      notable: [],
    });
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("finance");
    const sd = subdomainByPath(resp, "finance/newteam");
    expect(sd).toBeDefined();
    expect(sd?.name).toBe("newteam");
    expect(sd?.description).toBe("New team workspace");
    expect(subdomainByPath(resp, "finance/ap")).toBeDefined();
  });

  // spec: §4.5.2 — a workspace-local DOMAIN.md include: resolves over the merged
  // view (registry catalog ∪ overlay), pulling in both a registry artifact and
  // an overlay artifact while excluding an out-of-scope one. The fetcher serves
  // both /v1/load_domain and /v1/catalog by inspecting the URL.
  it("resolves include over the merged registry-plus-overlay view", async () => {
    await overlayArtifact(overlay, "finance/ap/overlay-pay", { desc: "overlay pay" });
    await overlayDomain(overlay, "drafts", "---\ninclude:\n  - finance/ap/*\n---\n");
    const fetcher = registryFetcher(
      { path: "drafts", subdomains: [], notable: [] },
      [
        { id: "finance/ap/registry-pay", type: "skill", summary: "registry pay" },
        { id: "other/unrelated", type: "skill", summary: "nope" },
      ],
    );
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("drafts");
    const got = notableByID(resp);
    expect(got.has("finance/ap/registry-pay")).toBe(true);
    expect(got.has("finance/ap/overlay-pay")).toBe(true);
    expect(got.has("other/unrelated")).toBe(false);
  });

  // spec: §4.5.3 — an overlay DOMAIN.md unlisted: true removes the folder and
  // its subtree from the parent's enumeration.
  it("prunes an overlay-unlisted subdomain", async () => {
    await overlayDomain(overlay, "finance/secret", "---\nunlisted: true\n---\n");
    const fetcher = registryFetcher({
      path: "finance",
      subdomains: [
        { path: "finance/ap", name: "ap", description: "AP" },
        { path: "finance/secret", name: "secret", description: "Secret" },
      ],
      notable: [],
    });
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("finance");
    expect(subdomainByPath(resp, "finance/secret")).toBeUndefined();
    expect(subdomainByPath(resp, "finance/ap")).toBeDefined();
  });

  // spec: §4.5.5 — an overlay DOMAIN.md at a child path overrides that child's
  // short description (frontmatter description, never the body).
  it("overrides a child description from the overlay frontmatter", async () => {
    await overlayDomain(
      overlay,
      "finance/ap",
      "---\ndescription: Local AP overrides\n---\nbody never shown for a child\n",
    );
    const fetcher = registryFetcher({
      path: "finance",
      subdomains: [{ path: "finance/ap", name: "ap", description: "Registry AP" }],
      notable: [],
    });
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("finance");
    const sd = subdomainByPath(resp, "finance/ap");
    expect(sd?.description).toBe("Local AP overrides");
  });

  // spec: §4.5.2 / §6.4 — a domain that exists only in the workspace overlay is
  // part of the effective view; the registry 404s it (it never sees the
  // overlay), so the SDK synthesizes an empty result and composes the overlay
  // onto it. The local include: resolves over the merged view.
  it("resolves an overlay-only domain the registry 404s", async () => {
    await overlayDomain(overlay, "drafts", "---\ndescription: Local drafts\ninclude:\n  - finance/ap/*\n---\n");
    await overlayArtifact(overlay, "finance/ap/overlay-pay", { desc: "overlay pay" });
    const catalog = [{ id: "finance/ap/registry-pay", type: "skill", summary: "registry pay" }];
    const fetcher: typeof fetch = async (input) => {
      const url = new URL(String(input));
      if (url.pathname === "/v1/load_domain") {
        return new Response(JSON.stringify({ code: "domain.not_found", message: "not here" }), {
          status: 404,
        });
      }
      if (url.pathname === "/v1/catalog") {
        const scope = url.searchParams.get("scope") ?? "";
        const arts = catalog.filter((e) => scope === "" || e.id === scope || hasSegPrefix(e.id, scope));
        return new Response(JSON.stringify({ ids: arts.map((a) => a.id), artifacts: arts }), { status: 200 });
      }
      return new Response("not found", { status: 404 });
    };
    const c = new Client({ registry: "http://reg", overlayPath: overlay, fetcher });
    const resp = await c.loadDomain("drafts");
    const ids = notableByID(resp);
    expect(ids.has("finance/ap/registry-pay")).toBe(true); // registry artifact via merged-view catalog
    expect(ids.has("finance/ap/overlay-pay")).toBe(true); // overlay artifact
    expect(resp.path).toBe("drafts");
  });
});
